compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.cpp

```diff
@@ -7,201 +7,275 @@
+        auto maybeConstantStrideX = getConstantIntValue(stridesX[i]);
+        auto maybeConstantStrideY = getConstantIntValue(stridesY[i]);
+        if (maybeConstantOffset.has_value() &&
+            maybeConstantStrideX.has_value() &&
+            maybeConstantStrideY.has_value()) {
+          int64_t offset = maybeConstantOffset.value();
+          int64_t strideX = maybeConstantStrideX.value();
+          int64_t strideY = maybeConstantStrideY.value();
+          int64_t numerator = offset * strideX;
+          int64_t denominator = strideY;
+          if (numerator % denominator == 0) {
+            offsetsX[i] = getAsIndexOpFoldResult(ctx, numerator / denominator);
+            stridesX[i] = stridesY[i];
+          }
+        }
+      }
+    }
   }
+}
 
-  // Equality of the last N elements of the access patterns of A and B with N =
-  // min(sizeA, sizeB) results in some simple cases in which the access
-  // patterns are combinable. Note that abs(sizeB - sizeA) should <= 1 and this
-  // is checked for earlier, so just asserted here.
-  assert(std::abs((ssize_t)offsetsB.size() - (ssize_t)offsetsA.size()) <= 1 &&
-         "The distance between the indices should be smaller or equal to one.");
-  size_t indexA = offsetsA.size() > offsetsB.size() ? 1 : 0;
-  size_t indexB = offsetsB.size() > offsetsA.size() ? 1 : 0;
-  if (areAccessPatternsEqualFromIndices(offsetsA, sizesA, stridesA, offsetsB,
-                                        sizesB, stridesB, indexA, indexB)) {
-    if (offsetsA.size() == offsetsB.size()) {
-      return true;
-    } else if (offsetsA.size() > offsetsB.size()) {
-      // The access pattern A has N repetitions of access pattern B, so they can
-      // be combined together into N+1 repetitions.
-      return isConstantIntValue(stridesA[0], 0);
+/// Background info: offsets, sizes, and strides define an access pattern into
+/// an array, where the i'th element accessed, for 0 <= i < prod_{d<D} sizes[d],
+/// is at index
+///
+///     sum_{d<D} (l(d,i) + offset[d]) * stride[d]                (1)
+///
+/// where l(d,i) is the component of global index `i` in dimension `d`:
+///
+///     i  = sum_{d<D} l(d,i) * size[d]                           (2)
+///
+/// Equation (1) can be rewritten with a global offset as
+///
+///     global_offset + sum_{d<D} l(d,i) * stride[d]              (3)
+///
+/// where the global offset is
+///
+///     global_offset = sum_{d<D} offset[d] * stride[d].
+///
+/// This function computes the difference between the global offsets of two
+/// access patterns. If it is not constant, i.e. if the difference contains
+/// an MLIR value which is not a constant, then nullopt is returned.
+///
+/// This function is useful when determining if the access pattern A, followed
+/// by the access pattern B, can be merged into a single access pattern.
+///
+/// \return global_offset(X) - global_offset(Y).
+std::optional<int64_t> getGlobalOffsetDifference(
```

Comment:
I think it would be useful to add unit tests for this utility (and other similar utilities).

Reply:
I've added some in the .cpp test file for the constant cases. For the non-constant cases, I'm in the process of adding in the combine_strided_ops.mlir, need to add some negative cases though (non-constant global offset difference). Tomorrow..